core/org.openjdk.jmc.agent/src/main/java/org/openjdk/jmc/agent/Transformer.java

```diff
@@ -62,30 +64,42 @@ public Transformer(TransformRegistry registry) {
 		if (!registry.hasPendingTransforms(className)) {
 			return registry.isRevertIntrumentation() ? classfileBuffer : null;
 		}
-		return doTransforms(registry.getTransformData(className), classfileBuffer, loader, protectionDomain);
+
+		// load the class for reflective inspection if not loaded already
+		if (classBeingRedefined == null) {
+			try {
+				// Don't reuse this class loader instance. We want the loader and its class to unload after we're done.
+				classBeingRedefined = new InspectionClassLoader(loader)
```

Comment:
We only need to parse/load the class if we are looking up fields. Can't we do this later, and only if we need it lazily? It's seems a bit premature to do it here in the Transformer.

Reply:
@tabjy - did you have a chance to look at this?

Reply:
Sorry, I was on vacation earlier, and I'm now slowly catching with up things. I'm looking into this right now.

Reply:
Classloading with `InspectionClassLoader` is delayed to `JFRClassVisitor` and is only loaded when necessary (ie. when having field evaluations.)